### PR TITLE
[tuner] Add abstractmethod in autotune.py. Create punet_autotune.py

### PR DESCRIPTION
**Pull request by RattataKing**
- Rename variables/functions with "unet" in their names in `autotune.py` and `test_autotune.py`
- Add class `TuningClient` and `DefaultTuningClient` in `autotune.py`
- Add `punet_autotune.py`
- Edit dir and path names in existing bash scripts

This PR generalized tuner by providing a basic python class API, decoupled it from the specific model (i.e. punet). The final goal is to make tuner support models other than SDXL.

### CODE REVIEW

File: tuning/autotune.py [hunk 1]
```diff
@@ -60,35 +62,32 @@
 
 
 @dataclass
-class CandidateTracker:
+class CandidateTracker(ABC):
```
Comment (kuhar) [inline tuning/autotune.py:65]: Why does this inherit from ABC?

File: tuning/autotune.py [hunk 2]
```diff
@@ -142,12 +143,75 @@ def get_candidate_mlir_path(self, candidate_id: int) -> Path:
...
+    def get_model_compile_command(self, candidate_tracker) -> list[str]:
+        pass
+
+    @abstractmethod
+    def get_model_benchmark_command(self, candidate_tracker) -> list[str]:
+        pass
+
+    @abstractmethod
+    def get_compiled_dispatch_index(self, file_path: Path) -> int:
+        pass
```
Comment (kuhar) [inline tuning/autotune.py:174]: Why does the client decide the index?

File: tuning/autotune.py [hunk 3]
```diff
@@ -142,12 +143,75 @@ def get_candidate_mlir_path(self, candidate_id: int) -> Path:
...
+    def get_candidate_spec_filename(self, candidate_id: int) -> str:
+        pass
+
+    @abstractmethod
+    def get_compiled_model_index(self, file_path: Path) -> int:
+        pass
+
+
+@dataclass
+class DefaultTuningClient(TuningClient):
```
Comment (kuhar) [inline tuning/autotune.py:186]: Why do we need this?

File: tuning/autotune.py (hunk 1 above)
Comment (RattataKing) [reply tuning/autotune.py]: a mistake, removed

File: tuning/autotune.py (hunk 2 above)
Comment (RattataKing) [reply tuning/autotune.py]: originally wanted to give users more freedom on file naming format

File: tuning/autotune.py (hunk 3 above)
Comment (RattataKing) [reply tuning/autotune.py:183]: to pass mypy type checking

File: tuning/autotune.py [hunk 4]
```diff
@@ -142,11 +142,41 @@ def get_candidate_mlir_path(self, candidate_id: int) -> Path:
...
+        return int(file_path.stem)
+
+    def get_candidate_spec_filename(self, candidate_id: int) -> str:
+        return f"{candidate_id}_spec.mlir"
+
+    def get_compiled_model_index(self, file_path: Path) -> int:
+        return int(file_path.stem.split("_")[-1])
+
+
+@dataclass
```
Comment (kuhar) [inline tuning/autotune.py:160]: Why are we making this a data class?

File: tuning/punet_autotune.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import autotune
+from dataclasses import dataclass
+from pathlib import Path
+
+
+@dataclass
```
Comment (kuhar) [inline tuning/punet_autotune.py:12]: same here

**Review by kuhar — APPROVED** [review]
LGTM

**Pull request merged**